src/scib_metrics/utils/_lisi.py

```diff
@@ -0,0 +1,80 @@
+import itertools
+from dataclasses import dataclass
+from typing import Tuple, Union
+
+import jax
+import jax.numpy as jnp
+import numpy as np
+
+NdArray = Union[np.ndarray, jnp.ndarray]
+
+
+@jax.jit
+def Hbeta(knn_dists_row: jnp.ndarray, beta: float) -> Tuple[jnp.ndarray, jnp.ndarray]:
+    """
+    Helper function for simpson index computation
+    """
+    P = jnp.exp(-knn_dists_row * beta)
+    sumP = jnp.nansum(P)
+    H = jnp.where(sumP == 0, 0, jnp.log(sumP) + beta * jnp.nansum(knn_dists_row * P) / sumP)
+    P = jnp.where(sumP == 0, jnp.zeros_like(knn_dists_row), P / sumP)
+    return H, P
+
+
+@jax.jit
+def _get_neighbor_probability(knn_dists_row: jnp.ndarray, perplexity: float, tol: float) -> Tuple[jnp.ndarray, jnp.ndarray]:
+    beta = 1
+    betamin = -jnp.inf
+    betamax = jnp.inf
+    H, P = Hbeta(knn_dists_row, beta)
+    Hdiff = H - jnp.log(perplexity)
+
+    def _get_neighbor_probability_step(state):
+        _, _, Hdiff, beta, betamin, betamax, tries = state
+        new_betamin = jnp.where(Hdiff > 0, beta, betamin)
+        new_betamax = jnp.where(Hdiff > 0, betamax, beta)
+        new_beta = jnp.where(Hdiff > 0, jnp.where(betamax == jnp.inf, beta * 2, (beta + betamax) / 2),
+                             jnp.where(betamin == -jnp.inf, beta / 2, (beta + betamin) / 2))
+        new_H, new_P = Hbeta(knn_dists_row, beta)
+        new_Hdiff = new_H - jnp.log(perplexity)
+        return new_H, new_P, new_Hdiff, new_beta, new_betamin, new_betamax, tries + 1
```

Comment:
Maybe we should use the flax dataclass (e.g., their trainstate) to represent this

https://flax.readthedocs.io/en/latest/api_reference/flax.struct.html

see the replace method for how to update it inplace